sympy/matrices/determinant.py

```diff
@@ -352,17 +364,15 @@ def _charpoly(M, x='lambda', simplify=_simplify):
     >>> from sympy.abc import x, y
     >>> M = Matrix([[1, 3], [2, 0]])
     >>> M.charpoly()
-    PurePoly(lambda**2 - lambda - 6, lambda, domain='ZZ')
-    >>> M.charpoly(x) == M.charpoly(y)
-    True
+    Poly(x**2 - x - 6, x, domain='QQ')
```

Comment:
Why has it changed from `x` to `lambda` and from `PurePoly` to `Poly`?

Reply:
Since while returning we are using `Poly` now, instead of `PurePoly`.
If we don't provide any variable then it will use `x` as deafult before it used to be `lambda`, but in the test cases `x` was provided and while using `PurePoly`, the variable didn't matter, but now it matters. So either we can set the default as `lambda` or `x`.
Kindly give your opinion.
Thanks.

Reply:
> Since while returning we are using `Poly` now, instead of `PurePoly`.

Why has it changed from `PurePoly` to `Poly`?

Reply:
> Why has it changed from `PurePoly` to `Poly`?

Sorry about that.

The latest commit still gives assertion error for the case mentioned - [test_eigen_slow](https://github.com/sympy/sympy/pull/21037#issuecomment-792248227)